hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -0,0 +1,395 @@
+                unpark((Thread) waiter);
+                return;
+            } else if (waiter instanceof ExecutionCallback) {
+                unblock((ExecutionCallback) waiter, executor);
+                return;
+            } else if (waiter.getClass() == WaitNode.class) {
+                WaitNode waitNode = (WaitNode) waiter;
+                unblockAll(waitNode.waiter, waitNode.executor);
+                waiter = waitNode.next;
+            } else {
+                return;
+            }
+        }
+    }
+
+    private void unblock(final ExecutionCallback<V> callback, Executor executor) {
+        try {
+            executor.execute(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        Object value = resolve(state);
+                        if (value instanceof Throwable) {
+                            Throwable error = (Throwable) value;
+                            if (error instanceof ExecutionException) {
+                                error = error.getCause();
+                            }
+                            callback.onFailure(error);
+                        } else {
+                            callback.onResponse((V) value);
+                        }
+                    } catch (Throwable cause) {
+                        logger.severe("Failed asynchronous execution of execution callback: " + callback
+                                + "for call " + invocationToString(), cause);
+                    }
+                }
+            });
+        } catch (RejectedExecutionException e) {
+            logger.warning("Execution of callback: " + callback + " is rejected!", e);
+        }
+    }
+
+    protected abstract String invocationToString();
+
+    protected Object resolve(Object value) {
+        return value;
+    }
+
+    protected abstract V resolveAndThrow(Object state) throws ExecutionException, InterruptedException;
+
+    protected abstract TimeoutException newTimeoutException(long timeout, TimeUnit unit);
+
+    /**
+     * Registers a waiter (so a Thread that blocks on the future completion, or a {@link ExecutionCallback} notified
+     * a     *
+     *
+     * @param waiter   the waiter
+     * @param executor the {@link Executor} to use in case of an {@link ExecutionCallback}.
+     * @return VOID if the registration was a success, anything else but void is the response.
+     */
+    private Object registerWaiter(Object waiter, Executor executor) {
+        WaitNode waitNode = null;
+        for (; ; ) {
+            final Object oldState = state;
+            if (isDone(oldState)) {
+                return oldState;
+            }
+
+            Object newState;
+            if (oldState == VOID && (executor == null)) {
+                // Nothing is syncing on this future, so instead of creating a WaitNode, we just try
+                // the cas the thread (so no extra litter)
```

Comment:
Should this be `to cas the thread`? Otherwise I didn't get the meaning of the comment.